### ISSUE TITLE
Be able to pass the generic type as a type argument

### ISSUE THREAD

**Issue by kasamachenkaow** [microsoft/TypeScript#43735]
# Suggestion

## 🔍 Search Terms

generic type, passing generic type

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

To be able to pass the generic type as a type argument

## 📃 Motivating Example

Consider these statements
```ts
type BaseResult = {}
type BaseType<TResult> = {
   build: () => TResult
}

type BuildReturnType<U extends BaseType<BaseResult>, NResult> = ReturnType<U['build']> & NResult

type SubTypeA<TResult> = TypeBase<TResult> & {
    withX: <U extends SubTypeA<BaseResult>>(this: U) => SubTypeA<BuildReturnType<U, XResult>> 
    withY: <U extends SubTypeA<BaseResult>>(this: U) => SubTypeA<BuildReturnType<U, YResult>> 
    withZ: <U extends SubTypeA<BaseResult>>(this: U) => SubTypeA<BuildReturnType<U, ZResult>> 
}
```
as you will notice that on the line `withX`, `withY` and `withZ` they are almost exactly the same except for the passing their own result type into the `BuildReturnType<U>`

so it would be nice if we could extract the duplicated lines to be a reusable type like this
```ts
type WithN<TSubType extends BaseType, NResult> = <U extends TSubType<BaseResult>>(this: U) => TSubType<BuildReturnType<U, NResult>> 
```

and then those 3 lines would be refactored to be

```ts
type SubTypeA<TResult> = TypeBase<TResult> & {
    withX: WithN<SubTypeA, XResult>
    withY: WithN<SubTypeA, YResult>
    withZ: WithN<SubTypeA, ZResult>
}
```

unfortunately the above code will throw errors on those 3 lines and in the `WithN<>` implementaion
since the generic type always requires the type arguments (unless optional ones specified) to be resolved into non-generic type

basically we can't pass the generic type itself (without fill with any type parameters yet) as a type argument

so I would like to know the possibility of introducing that

## 💻 Use Cases

- To reduce the duplicate code when we want to refactor the type of generic functions as show in the example for better readability

**Comment by RyanCavanaugh** [comment]
Seems to be a duplicate of #1213
